src/main/java/org/waarp/openr66/protocol/networkhandler/ssl/NetworkSslServerInitializer.java

```diff
@@ -70,24 +70,24 @@ protected void initChannel(SocketChannel ch) throws Exception {
         }
         pipeline.addLast("ssl", sslHandler);
 
-        pipeline.addLast("codec", new NetworkPacketCodec());
         pipeline.addLast(NetworkServerInitializer.TIMEOUT,
-                new IdleStateHandler(0, 0, Configuration.configuration.getTIMEOUTCON(), TimeUnit.MILLISECONDS));
-        GlobalTrafficHandler handler = Configuration.configuration
-                .getGlobalTrafficShapingHandler();
-        if (handler != null) {
-            pipeline.addLast(NetworkServerInitializer.LIMIT, handler);
-        }
-        ChannelTrafficShapingHandler trafficChannel = null;
-        try {
-            trafficChannel =
-                    Configuration.configuration
-                            .newChannelTrafficShapingHandler();
-            pipeline.addLast(NetworkServerInitializer.LIMITCHANNEL, trafficChannel);
-        } catch (OpenR66ProtocolNoDataException e) {
-        }
-        pipeline.addLast(Configuration.configuration.getHandlerGroup(), "handler", new NetworkSslServerHandler(
-                !this.isClient));
+                new IdleStateHandler(0, 0,
+                    Configuration.configuration.getTIMEOUTCON(),
+                    TimeUnit.MILLISECONDS));
+
+        // Global limitation
+        pipeline.addLast(NetworkServerInitializer.LIMITGLOBAL,
+                Configuration.configuration.getGlobalTrafficShapingHandler());
+        // Per channel limitation
+        pipeline.addLast(NetworkServerInitializer.LIMITCHANNEL,
+                new ChannelTrafficShapingHandler(
+                    Configuration.configuration.getServerChannelWriteLimit(),
+                    Configuration.configuration.getServerChannelReadLimit(),
+                    Configuration.configuration.getDelayLimit()));
+ 
+        pipeline.addLast("codec", new NetworkPacketCodec());
```

Comment:
Does this should not be before any TrafficShaping ? not so sure on my side... Maybe it is the right position there...

Reply:
Putting the TrafficShapinghandlers before the Codec seems a better way to organize the channel.

The previous configuration limited the Codec speed and not the network writing/reading speed. The Codec was fed bytes as the pace imposed by the TrafficShapinghandlers but was writing on the network as fast as the connection allowed it.

The previous pipeline forced the use of custom TrafficShapinghandlers to reflect the proper size of the network traffic. (It just added 13 bytes to the traffic count to take account of the overhead added by the NetworkCodec).

This pipeline  configuration is cleaner on the traffic shaping, the traffic computation, and allows the use standard netty ShapingHandler.

Reply:
perfect